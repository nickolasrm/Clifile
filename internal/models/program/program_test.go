package program_test

import (
	"testing"

	"github.com/nickolasrm/clifile/internal/models/call"
	"github.com/nickolasrm/clifile/internal/models/program"
	"github.com/nickolasrm/clifile/internal/models/rule"
	"github.com/nickolasrm/clifile/internal/models/variable"
	. "github.com/nickolasrm/clifile/test/util"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

func TestProgram(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Program Suite")
}

var _ = Describe("Program", func() {
	When("a new program is created", func() {
		var p *program.Program

		BeforeEach(func() {
			p = program.NewProgram()
		})

		It("should return a program type", func() {
			MatchSnapshot(p)
		})

		When("Doc() is called", func() {
			It("should return the program doc", func() {
				MatchSnapshot(p.Doc())
			})
		})
		When("SetDoc() is called", func() {
			It("should replace the program doc", func() {
				p.SetDoc("new doc")
				MatchSnapshot(p)
			})
		})
		When("Call() is called", func() {
			When("the passed call exists", func() {
				It("should return the call", func() {
					c := call.NewCall("call", "fn", map[string]string{})
					p.AddCall(c)
					MatchSnapshot(p.Call("call"))
				})
			})
			When("the passed call does not exist", func() {
				It("should return nil", func() {
					Expect(p.Call("call")).To(BeNil())
				})
			})
		})
		When("AddCall() is called", func() {
			It("should add the call to the program", func() {
				c := call.NewCall("call", "fn", map[string]string{})
				p.AddCall(c)
				MatchSnapshot(p)
			})
		})
		When("Rules() is called", func() {
			When("the program has no rules", func() {
				It("should return an empty slice", func() {
					Expect(p.Rules()).To(BeEmpty())
				})
			})
			When("the program has rules", func() {
				It("should return the rules", func() {
					r := rule.NewRule("rule", []string{"pos1"}, "doc", "action")
					p.AddRule(r)
					MatchSnapshot(p.Rules())
				})
			})
		})
		When("AddRule() is called", func() {
			It("should add the rule to the program", func() {
				r := rule.NewRule("rule", []string{"pos1"}, "doc", "action")
				p.AddRule(r)
				MatchSnapshot(p)
			})
		})
		When("Rule() is called", func() {
			When("the passed rule exists", func() {
				It("should return the rule", func() {
					r := rule.NewRule("rule", []string{"pos1"}, "doc", "action")
					p.AddRule(r)
					MatchSnapshot(p.Rule("rule"))
				})
			})
			When("the passed rule does not exist", func() {
				It("should return nil", func() {
					Expect(p.Rule("rule")).To(BeNil())
				})
			})
		})
		When("AddVariable() is called", func() {
			It("should add the variable to the program", func() {
				v := variable.NewVariable("var", "value")
				p.AddVariable(v)
				MatchSnapshot(p)
			})
		})
		When("Variable() is called", func() {
			When("the passed variable exists", func() {
				It("should return the variable", func() {
					v := variable.NewVariable("var", "value")
					p.AddVariable(v)
					MatchSnapshot(p.Variable("var"))
				})
			})
			When("the passed variable does not exist", func() {
				It("should return nil", func() {
					Expect(p.Variable("var")).To(BeNil())
				})
			})
		})
		When("SetVariable() is called", func() {
			When("the passed variable exists", func() {
				It("should set the variable value", func() {
					v := variable.NewVariable("var", "value")
					p.AddVariable(v)
					p.SetVariable("var", "new value")
					MatchSnapshot(p)
				})
			})
			When("the passed variable does not exist", func() {
				It("should create the variable", func() {
					p.SetVariable("var", "value")
					MatchSnapshot(p)
				})
			})
		})
	})
})
