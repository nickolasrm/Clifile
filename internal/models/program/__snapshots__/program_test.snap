
[Program when a new program is created should return a program type - 1]
[]interface {}{
    &program.Program{
        doc:       "Software Command Line Interface (CLI)\nUse this as shortcut for user-defined commands",
        variables: {
        },
        calls: {
        },
        rules: {
        },
    },
}
---

[Program when a new program is created when Doc() is called should return the program doc - 1]
[]interface {}{
    "Software Command Line Interface (CLI)\nUse this as shortcut for user-defined commands",
}
---

[Program when a new program is created when SetDoc() is called should replace the program doc - 1]
[]interface {}{
    &program.Program{
        doc:       "new doc",
        variables: {
        },
        calls: {
        },
        rules: {
        },
    },
}
---

[Program when a new program is created when Call() is called when the passed call exists should return the call - 1]
[]interface {}{
    &call.Call{
        name:      "call",
        function:  "fn",
        arguments: {},
    },
}
---

[Program when a new program is created when AddCall() is called should add the call to the program - 1]
[]interface {}{
    &program.Program{
        doc:       "Software Command Line Interface (CLI)\nUse this as shortcut for user-defined commands",
        variables: {
        },
        calls: {
            "call": &call.Call{
                name:      "call",
                function:  "fn",
                arguments: {},
            },
        },
        rules: {
        },
    },
}
---

[Program when a new program is created when Rules() is called when the program has rules should return the rules - 1]
[]interface {}{
    []*rule.Rule{
        &rule.Rule{
            name:       "rule",
            positional: {"pos1"},
            doc:        "doc",
            actions:    "action\n",
            rules:      {
            },
        },
    },
}
---

[Program when a new program is created when AddRule() is called should add the rule to the program - 1]
[]interface {}{
    &program.Program{
        doc:       "Software Command Line Interface (CLI)\nUse this as shortcut for user-defined commands",
        variables: {
        },
        calls: {
        },
        rules: {
            "rule": &rule.Rule{
                name:       "rule",
                positional: {"pos1"},
                doc:        "doc",
                actions:    "action\n",
                rules:      {
                },
            },
        },
    },
}
---

[Program when a new program is created when Rule() is called when the passed rule exists should return the rule - 1]
[]interface {}{
    &rule.Rule{
        name:       "rule",
        positional: {"pos1"},
        doc:        "doc",
        actions:    "action\n",
        rules:      {
        },
    },
}
---

[Program when a new program is created when AddVariable() is called should add the variable to the program - 1]
[]interface {}{
    &program.Program{
        doc:       "Software Command Line Interface (CLI)\nUse this as shortcut for user-defined commands",
        variables: {
            "var": &variable.Variable{name:"var", value:"value"},
        },
        calls: {
        },
        rules: {
        },
    },
}
---

[Program when a new program is created when Variable() is called when the passed variable exists should return the variable - 1]
[]interface {}{
    &variable.Variable{name:"var", value:"value"},
}
---

[Program when a new program is created when SetVariable() is called when the passed variable exists should set the variable value - 1]
[]interface {}{
    &program.Program{
        doc:       "Software Command Line Interface (CLI)\nUse this as shortcut for user-defined commands",
        variables: {
            "var": &variable.Variable{name:"var", value:"new value"},
        },
        calls: {
        },
        rules: {
        },
    },
}
---

[Program when a new program is created when SetVariable() is called when the passed variable does not exist should create the variable - 1]
[]interface {}{
    &program.Program{
        doc:       "Software Command Line Interface (CLI)\nUse this as shortcut for user-defined commands",
        variables: {
            "var": &variable.Variable{name:"var", value:"value"},
        },
        calls: {
        },
        rules: {
        },
    },
}
---
